census/src/main/java/io/grpc/census/CensusStatsModule.java

```diff
@@ -477,42 +461,58 @@ public ClientStreamTracer newClientStreamTracer(StreamInfo info, Metadata metada
       } else {
         attemptsPerCall.incrementAndGet();
       }
-      return tracer;
+      return new ClientTracer(this, module, parentCtx, startCtx, info);
     }
 
     // Called whenever each attempt is ended.
     void attemptEnded() {
-      if (activeStreams.decrementAndGet() == 0) {
-        // Race condition between two extremely close events does not matter because the difference
-        // in the result would be very small.
-        long lastInactiveTimeStamp =
-            this.lastInactiveTimeStamp.getAndSet(stopwatch.elapsed(TimeUnit.NANOSECONDS));
-        retryDelayNanos.addAndGet(-lastInactiveTimeStamp);
+      if (!module.recordFinishedRpcs) {
+        return;
+      }
+      boolean shouldRecordFinishedCall = false;
+      synchronized (lock) {
+        if (--activeStreams == 0) {
+          stopwatch.start();
+          if (callEnded && !finishedCallToBeRecorded) {
+            shouldRecordFinishedCall = true;
+            finishedCallToBeRecorded = true;
+          }
+        }
+      }
+      if (shouldRecordFinishedCall) {
+        recordFinishedCall();
       }
     }
 
     void callEnded(Status status) {
-      if (callEndedUpdater != null) {
-        if (callEndedUpdater.getAndSet(this, 1) != 0) {
+      if (!module.recordFinishedRpcs) {
+        return;
+      }
+      this.status = status;
+      boolean shouldRecordFinishedCall = false;
+      synchronized (lock) {
+        if (callEnded) {
```

Comment:
It's not possible unless there's some bug like #7921.

Reply:
Then how about we log a warning if it happens? An assert would be great, but I don't think we are that confident. A warning with "this is a bug" at least would mean we'd learn if it happens. If it does trigger, it seems it'd probably be at a low rate.

Reply:
Added logging.

Reply:
Uhh... if we think it can happen, then let's just have the notice of the bug here and return like you had.

Reply:
No, I mean delete the log statement. I was suggesting the log statement only if we thought it wasn't possible but it would be worrisome to add an assert. If we think it is possible, just having a comment seems good.

Reply:
Done.